runtime/fiber.c

```diff
@@ -1117,3 +1133,318 @@ CAMLexport void caml_raise_unhandled_effect (value effect)
 {
   caml_raise(caml_make_unhandled_effect_exn(effect));
 }
+
+/**** Dynamic Binding ****/
+
+/* Implementation of primitives to support Dynamic.t */
+
+extern value caml_fresh_oo_id(value v);
+
+/* Making a dynamic value */
+
+CAMLprim value caml_dynamic_make(value val)
+{
+  CAMLparam1(val);
+  /* TODO: consider other hash functions. This one is ~unique, which is nice */
+  value hash = caml_fresh_oo_id(Val_unit);
```

Comment:
Is there any way to give it a stable hash without adding an indirection to cache lookup? 
Alternatively could we use its address as the cache hash (but not the table hash) and just lose track of the cache entry when it moves? Or clear the entry when moved?

Reply:
There are various possibilities here, but I am deliberately punting performance hacks to later work, and focusing here on getting the semantics right. The same goes for @stedolan's notion to use a vector unit hack for the cache check.
If we used some bits of the address as the hash, then on a minor GC we'd not only have to flush young items from the per-thread cache, but also rehash all the per-thread root-binding tables (which use the same hash value).